File: DynamicKey/AgoraDynamicKey/deno/test/EducationTokenBuilderTest.js
/**
 * run this test with command:
 * deno test test/EducationTokenBuilderTest.js
 */
import { EducationTokenBuilder } from '../src/EducationTokenBuilder.js'
import { AccessToken2, kApaasServiceType } from '../src/AccessToken2.js'
import { assert } from 'https://deno.land/std/testing/asserts.ts'
const appId = '970CA35de60c44645bbae8a215061b33'
const appCertificate = '5CFd2fd1755d40ecb72977518be15d3b'
const expire = 600
const roomUuid = '123'
const userUuid = '2882341273'
const role = 1

// Verifies Education room-user token generation and parsing.
Deno.test('BuildRoomUserToken_Test', () => {
    let accessToken = new AccessToken2('', '', 0, 0)
    let token = EducationTokenBuilder.buildRoomUserToken(appId, appCertificate, roomUuid, userUuid, role, expire)
    accessToken.from_string(token)
    const service = accessToken.getServices(kApaasServiceType)[0]
    assert(appId === accessToken.appId)
    assert(expire === accessToken.expire)
    assert(roomUuid === service.__room_uuid)
    assert(userUuid === service.__user_uuid)
    assert(role === service.__role)
})

// Verifies Education user token generation and parsing.
Deno.test('BuildUserToken_Test', () => {
    let accessToken = new AccessToken2('', '', 0, 0)
    let token = EducationTokenBuilder.buildUserToken(appId, appCertificate, userUuid, expire)
    accessToken.from_string(token)
    const service = accessToken.getServices(kApaasServiceType)[0]

    assert(appId === accessToken.appId)
    assert(expire === accessToken.expire)
    assert(userUuid === service.__user_uuid)
})

// Verifies Education application token generation and parsing.
Deno.test('BuildAppToken_Test', () => {
    let accessToken = new AccessToken2('', '', 0, 0)
    let token = EducationTokenBuilder.buildAppToken(appId, appCertificate, expire)
    accessToken.from_string(token)
    assert(appId === accessToken.appId)
    assert(expire === accessToken.expire)
})

File: DynamicKey/AgoraDynamicKey/go/src/educationtokenbuilder/educationtokenbuilder_test.go
package educationtokenbuilder

import (
	"testing"

	accesstoken "github.com/AgoraIO/Tools/DynamicKey/AgoraDynamicKey/go/src/accesstoken2"
)

const (
	DataMockAppCertificate = "5CFd2fd1755d40ecb72977518be15d3b"
	DataMockAppId          = "970CA35de60c44645bbae8a215061b33"
	DataMockRoomUuid       = "123"
	DataMockUserUuid       = "2882341273"
	DataMockRole           = int16(1)
	DataMockExpire         = uint32(600)
)

func Test_BuildRoomUserToken(t *testing.T) {
	token, err := BuildRoomUserToken(DataMockAppId, DataMockAppCertificate, DataMockRoomUuid, DataMockUserUuid, DataMockRole, DataMockExpire)
	accesstoken.AssertNil(t, err)

	accessToken := accesstoken.CreateAccessToken()
	accessToken.Parse(token)

	accesstoken.AssertEqual(t, DataMockAppId, accessToken.AppId)
	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Expire)
	accesstoken.AssertEqual(t, true, accessToken.Services[accesstoken.ServiceTypeEducation] != nil)
	accesstoken.AssertEqual(t, DataMockRoomUuid, accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).RoomUuid)
	accesstoken.AssertEqual(t, DataMockUserUuid, accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).UserUuid)
	accesstoken.AssertEqual(t, DataMockRole, accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).Role)

	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).Privileges[accesstoken.PrivilegeEducationRoomUser])
	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Services[accesstoken.ServiceTypeRtm].(*accesstoken.ServiceRtm).Privileges[accesstoken.PrivilegeLogin])
	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Services[accesstoken.ServiceTypeChat].(*accesstoken.ServiceChat).Privileges[accesstoken.PrivilegeChatUser])
}

func Test_BuildUserToken(t *testing.T) {
	token, err := BuildUserToken(DataMockAppId, DataMockAppCertificate, DataMockUserUuid, DataMockExpire)
	accesstoken.AssertNil(t, err)

	accessToken := accesstoken.CreateAccessToken()
	accessToken.Parse(token)

	accesstoken.AssertEqual(t, DataMockAppId, accessToken.AppId)
	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Expire)
	accesstoken.AssertEqual(t, true, accessToken.Services[accesstoken.ServiceTypeEducation] != nil)

	accesstoken.AssertEqual(t, DataMockUserUuid, accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).UserUuid)
	accesstoken.AssertEqual(t, "", accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).RoomUuid)
	accesstoken.AssertEqual(t, int16(-1), accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).Role)

	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).Privileges[accesstoken.PrivilegeEducationUser])
}

func Test_BuildAppToken(t *testing.T) {
	token, err := BuildAppToken(DataMockAppId, DataMockAppCertificate, DataMockExpire)
	accesstoken.AssertNil(t, err)

	accessToken := accesstoken.CreateAccessToken()
	accessToken.Parse(token)

	accesstoken.AssertEqual(t, DataMockAppId, accessToken.AppId)
	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Expire)
	accesstoken.AssertEqual(t, true, accessToken.Services[accesstoken.ServiceTypeEducation] != nil)

	accesstoken.AssertEqual(t, "", accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).UserUuid)
	accesstoken.AssertEqual(t, "", accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).RoomUuid)
	accesstoken.AssertEqual(t, int16(-1), accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).Role)

	accesstoken.AssertEqual(t, DataMockExpire, accessToken.Services[accesstoken.ServiceTypeEducation].(*accesstoken.ServiceEducation).Privileges[accesstoken.PrivilegeEducationApp])
}
